test/discr/discr_ops_test.py

```diff
@@ -0,0 +1,287 @@
+                                zero_padding=True)
+
+    # partial derivative
+    par_div_f2 = par_div(f)
+
+    assert not par_div_f1 == par_div_f2
+
+
+def ndvolume(lin_size, ndim, dtype=np.float64):
+    """Hypercube phantom.
+
+    Parameters
+    ----------
+    lin_size : `int`
+       Size of array in each dimension
+    ndim : `int`
+        Number of dimensions
+    dtype : dtype
+        The type of the output array
+
+    """
+    s = [1]
+    vol = np.arange(lin_size, dtype=dtype)
+    for _ in range(ndim - 1):
+        s.insert(0, lin_size)
+        vol = vol * vol.reshape(s)
+    return vol
+
+
+def test_discrete_gradient():
+    """Discretized spatial gradient operator."""
+
+    discr_space = Rn(1)
+    with pytest.raises(TypeError):
+        DiscreteGradient(discr_space)
+
+    for ndim in range(1, 6):
+        # DiscreteLp Vector
+        vsize = 3
+        intvl = IntervalProd([0.] * ndim, [vsize] * ndim)
+        space = FunctionSpace(intvl)
+        discr_space = uniform_discr_fromspace(space, [vsize] * ndim)
+        dom_vec = discr_space.element(ndvolume(vsize, ndim))
+
+        # Gradient
+        grad = DiscreteGradient(discr_space)
+        grad(dom_vec)
+
+    # DiscreteLp Vector
+    ndim = 2
+    vsize = 3
+    intvl = IntervalProd([0.] * ndim, [vsize] * ndim)
+    space = FunctionSpace(intvl)
+    discr_space = uniform_discr_fromspace(space, [vsize] * ndim)
+    dom_vec = discr_space.element(ndvolume(vsize, ndim))
+
+    # Gradient
+    grad = DiscreteGradient(discr_space)
+    grad_vec = grad(dom_vec)
+    assert len(grad_vec) is ndim
+
+    # Adjoint operator
+    ran_vec = grad.range.element(
+        [ndvolume(vsize, ndim) ** 2] * ndim)
+    adj = grad.adjoint
+    adj_vec = adj(ran_vec)
+    lhs = ran_vec.inner(grad_vec)
+    rhs = dom_vec.inner(adj_vec)
+    assert not lhs == 0
+    assert not rhs == 0
+    assert lhs == rhs
+
+
+def test_discrete_divergence():
+    """Discretized spatial divergence operator."""
+
+    # Invalid arguments
+    discr_space = Rn(1)
+    with pytest.raises(TypeError):
+        DiscreteDivergence(discr_space)
+
+    # Check result of operator with explicit summation
+    data = np.array([[0., 1., 2., 3., 4.],
+                     [1., 2., 3., 4., 5.],
+                     [2., 3., 4., 5., 6.]])
+
+    # DiscreteLp
+    space = FunctionSpace(Rectangle([0, 0], [6, 2.5]))
+    discr_space = uniform_discr_fromspace(space, data.shape)
+
+    # Operator instance
+    div = DiscreteDivergence(discr_space)
+
+    # Apply operator
+    dom_vec = div.domain.element([data, data])
+    div_dom_vec = div(dom_vec)
+
+    # computation of divergence with helper function
+    dx0, dx1 = discr_space.grid.stride
+    df0 = finite_diff(data, axis=0, dx=dx0, zero_padding=True, edge_order=2)
+    df1 = finite_diff(data, axis=1, dx=dx1, zero_padding=True, edge_order=2)
+
+    assert all_equal(df0 + df1, div_dom_vec.asarray())
```

Comment:
No it is not actually. Nice catch. It is line 93 in `testutils` and the error is because `df0 + df1` is `2 x n` while `div_dom_vec` is `2n`. We need to try catch the statement `return iter1 == iter2` and return `false` on catch.